test_provider.py

```diff
@@ -0,0 +1,454 @@
+    def strip_constraints(constraint: str):
+        return constraint.split('/')[0].split('%')[0]
+
+    @staticmethod
+    def parse_magnitude(json: dict) -> Magnitude:
+        band = MagnitudeBands[json['name']]
+        value = json['value']
+        error = None
+        return Magnitude(band, value, error)
+    
+    @staticmethod
+    def get_program_dates(program_type: ProgramTypes, id: str, notes: list) -> Tuple[datetime, datetime]:
+        start = end = None
+        year = int(id[3:7])
+        next_year = year + 1
+        semester = id[7]
+        if program_type is ProgramTypes.FT:
+
+            for note in notes:
+                if 'title' in note.keys():
+                    title = note['title'].lower()
+                    if 'cycle' in title or 'active' in title:
+                        fields = title.strip().split(' ')
+                        months = []
+                        for field in fields:
+                            if '-' in field:
+                                months = field.split('-')
+                                if len(months) == 3:
+                                    break
+                        if len(months) == 0:
+                            for field in fields:
+                                f = field.lower().strip(' ,')
+                                if f in ['jan', 'feb', 'mar', 'apr', 'may', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec']:
+                                    months.append(f)
+                        im1 = months[0]
+                        im2 = months[-1]
+                        if semester == 'B' and im1 < 6:
+                            start = datetime(next_year, im1, 1)
+                            end = datetime(next_year, im2, 1)
+                        else:
+                            start = datetime(year, im1, 1)
+                            end = datetime(year, im2, 1) if im2 > im1 else datetime(next_year, im2, 1)
+                        break
+        else:
+            if semester == 'A':
+                start = datetime(int(year), 2, 1)
+                end = datetime(int(year), 7, 31)
+            else:
+                start = datetime(int(year), 8, 1)
+                end = datetime(int(next_year), 1, 31)
+        return start, end
+
+    def parse_timing_window(self, json: dict) -> TimingWindow:
+        tw_arr = []
+        for timing_window in json[JsonProvider._TimingWindowsKeys.TIMING_WINDOWS.value]:
+            tw = TimingWindow()
+            repeat = JsonProvider._TimingWindowsKeys.REPEAT.value
+            tw.start = datetime.fromtimestamp(timing_window[JsonProvider._TimingWindowsKeys.START.value]/1000)
+            tw.duration = timedelta(milliseconds=timing_window[JsonProvider._TimingWindowsKeys.DURATION.value])
+            tw.repeat = timing_window[repeat]
+            tw.period = timedelta(milliseconds=timing_window[repeat]) if timing_window[repeat] > 0 else None
+            tw_arr.append(tw)
+        return tw_arr
```

Comment:
We want to retrieve the data and THEN create the `dataclass`:
```
repeat = ...
start = ...
duration = ...
...
tw = TimingWindow(...) # put fields in here
tw_arr.append(tw)
```

Dataclasses should be considered mutable and we should try to only populate them at the time they are constructed: we should not be constructing them and then populating them.

Reply:
I had to change this around: I intended this method to parse out a single `TimingWindow` and not the entire list.